### PR TITLE
refactor: remove state_override from set_scheme

### PR DESCRIPTION
**Pull request by anonrig**
Fixes https://github.com/ada-url/ada/issues/75 and https://github.com/ada-url/ada/issues/76

### CODE REVIEW

File: src/helpers.cpp [hunk 1]
```diff
@@ -79,10 +79,31 @@ namespace ada::helpers {
...
+        /**
+         * TODO: Ok. So if we arrive here then view has an unclosed [,
+         * Is the URL valid???
+         */
+      } else {
+        location = url.is_special() ? view.find_first_of(":[/?\\#", location) : view.find_first_of(":[/?#", location);
+      }
+    }
+
+    view = std::string_view(view.data(), (location != std::string_view::npos) ? location : view.size());
```
Comment (lemire) [inline src/helpers.cpp:99]: Is this line needed?

File: src/implementation.cpp [hunk 2]
```diff
@@ -24,13 +24,12 @@ namespace ada {
 
   ada_warn_unused url parse(std::string_view input,
                             std::optional<ada::url> base_url,
-                            ada::encoding_type encoding,
-                            std::optional<ada::state> state) noexcept {
+                            ada::encoding_type encoding) noexcept {
```
Comment (lemire) [inline src/implementation.cpp:27]: I think that the `noexcept` are unsafe because we use strings that might throw.

**Review by lemire — APPROVED** [review]
This should be performance neutral, but it will enable better tests and performance later.

**Comment by lemire** [comment]
Wherever a std::string is used, we should remove noexcept. Wherever we call a function that is not noexcept, we should remove the noexcept. Otherwise we risk having a TERMINATE. This is a minor point since an exception is probably fatal for us anyhow (almost surely it means we have run out of memory), but still... some people might object.

**Comment by lemire** [comment]
You may need to sync with the main branch.

File: include/ada/url.h [hunk 3]
```diff
@@ -228,6 +228,11 @@ namespace ada {
      */
     ada_really_inline bool parse_prepared_path(const std::string_view input);
 
+    /**
+     * Parse scheme assuming it ends with ':'
```
Comment (lemire) [inline include/ada/url.h:232]: This new condition (assuming it ends with ':') is a bit odd.

Comment (anonrig) [reply include/ada/url.h]: It was wrong. I updated it.

File: src/helpers.cpp (hunk 1 above)
Comment (anonrig) [reply src/helpers.cpp]: I changed this line to remove_suffix. It's making things complicated if we don't do it here. I'll be happy to take a look into this in a different pull request

**Comment by anonrig** [comment]
I've applied the changes requested. I'm merging this pull request 👍

**Pull request merged**